intellij/src/saros/intellij/ui/tree/ContactTreeRootNode.java

```diff
@@ -3,111 +3,108 @@
-import java.util.Collection;
 import java.util.HashMap;
 import java.util.Map;
+import java.util.Optional;
 import javax.swing.tree.DefaultMutableTreeNode;
 import javax.swing.tree.DefaultTreeModel;
-import javax.swing.tree.MutableTreeNode;
-import org.jivesoftware.smack.Roster;
-import org.jivesoftware.smack.RosterEntry;
-import org.jivesoftware.smack.packet.Presence;
 import saros.SarosPluginContext;
 import saros.intellij.ui.util.IconManager;
 import saros.intellij.ui.views.SarosMainPanelView;
 import saros.net.xmpp.JID;
-import saros.net.xmpp.roster.IRosterListener;
-import saros.net.xmpp.roster.RosterTracker;
+import saros.net.xmpp.contact.IContactsUpdate;
+import saros.net.xmpp.contact.XMPPContact;
+import saros.net.xmpp.contact.XMPPContactsService;
 import saros.repackaged.picocontainer.annotations.Inject;
 
 /**
  * Root node for the tree of contacts, that a user has in their contact list.
  *
- * <p>It registers as {@link IRosterListener} and keeps the displayed contacts in sync with the
- * server.
- *
  * <p><b>NOTE:</b>This component and any component added here must be correctly torn down when the
  * project the components belong to is closed. See {@link SarosMainPanelView}.
  */
-
-/*
- * TODO the visual representation of the roster has
- * the same issue as the Eclipse counterpart, it cannot
- * handle multiple JIDs with different resources
- */
-public class ContactTreeRootNode extends DefaultMutableTreeNode
-    implements IRosterListener, Disposable {
-
+public class ContactTreeRootNode extends DefaultMutableTreeNode implements Disposable {
   private static final long serialVersionUID = 1L;
-
-  public static final String TREE_TITLE = "Contacts";
-
-  private final SessionAndContactsTreeView treeView;
-
-  private final Map<JID, ContactInfo> visibleContactsMap = new HashMap<>();
-  private final Map<JID, ContactInfo> hiddenContactsMap = new HashMap<>();
+  private static final String TREE_TITLE = "Contacts";
 
   private final DefaultTreeModel treeModel;
+  private final SessionAndContactsTreeView treeView;
 
-  @Inject private RosterTracker rosterTracker;
-
-  private Roster roster;
+  private final Map<XMPPContact, ContactInfo> visibleContactsMap = new HashMap<>();
+  private final Map<XMPPContact, ContactInfo> hiddenContactsMap = new HashMap<>();
+
+  @Inject private XMPPContactsService contactsService;
+
+  private final IContactsUpdate contactsUpdate =
+      new IContactsUpdate() {
+
+        @Override
+        public void update(Optional<XMPPContact> contact, UpdateType updateType) {
+          UIUtil.invokeLaterIfNeeded(
+              () -> {
+                if (updateType == UpdateType.ADDED) {
+                  contact.ifPresent(ContactTreeRootNode.this::addContact);
+                } else if (updateType == UpdateType.CONNECTED) {
+                  contactsService
+                      .getAllContacts()
+                      .stream()
+                      .forEach(ContactTreeRootNode.this::addContact);
```

Comment:
I don't really understand this case. From reading the logic, it seems to me like it is re-adding all contacts whenever an account connects to the XMPP server. This should lead to each account entry being present multiple times as I don't see a call that cleans up the list.

But this is not the case. Every entry is always only present once.

Could you explain this to me? Where is my mistake in reading the logic? What exactly does this logic do and which case does it cover?

Reply:
Good that you checked this :D

First I reimplemented the `IRosterListener` (rosterChanged call) here. Split it to not_connected and connected which introduced this part and than I didn't check if I could delete it.

The old impl looks like it did the same... But I removed it.

Reply:
And to answer why this doesn't duplicate contacts:
https://github.com/saros-project/saros/blob/9cba0106eaf9eb023aa4dbc0b9f0979c793052d9/intellij/src/saros/intellij/ui/tree/ContactTreeRootNode.java#L87-L89
So instead of creating a new contact the code stops here.